### PR TITLE
script.copacetic.helper 1.0.0

### PR DESCRIPTION
**Pull request by realcopacetic**
Initial repo submission for script.copacetic.helper

### Description
A new addon dependency for the skin Copacetic, submitted to rep-skins repo. This addon provides script, service and plugin modules for use by Copacetic.

### Checklist:
- [x] My code follows the [add-on rules](http://kodi.wiki/view/Add-on_rules) and [piracy stance](http://kodi.wiki/view/Official:Forum_rules#Piracy_Policy) of this project. 
- [X] I have read the [CONTRIBUTING](https://github.com/xbmc/repo-scripts/blob/master/CONTRIBUTING.md) document
- [X] Each add-on submission should be a single commit with using the following style: [script.foo.bar] 1.0.0

Additional information :
- Submitting your add-on to this specific branch makes it available to any Kodi version equal or higher than the branch name with the applicable Kodi dependencies limits.
- [add-on development](http://kodi.wiki/view/Add-on_development) wiki page.
- Kodi [pydocs](http://kodi.wiki/view/PyDocs) provide information about the Python API
- [PEP8](https://www.python.org/dev/peps/pep-0008/) codingstyle which is considered best practise but not mandatory.
- This add-on repository has automated code guideline check which could help you improve your coding. You can find the results of these check at [Codacy](https://www.codacy.com/app/Kodi/repo-scripts/dashboard). You can create your own account as well to continuously monitor your python coding before submitting to repo.
- Development questions can be asked in the [add-on development](http://forum.kodi.tv/forumdisplay.php?fid=26) section on the Kodi forum.

### CODE REVIEW

**Comment by Lunatixz** [comment]
How does this script differ from script.embuary.helper and plugin.video.themoviedb.helper?

**Comment by realcopacetic** [comment]
> How does this script differ from script.embuary.helper and plugin.video.themoviedb.helper?

Hi, it started as an exercise for me to learn python by reverse engineering the functionality I needed from those two you mentioned, but expanded in scope to include a number of new functions.  Below I have added the functions/methods I wrote for this script to be used by skin.copacetic that weren't already provided by those two you mentioned.

For scripts:
- function to clean filenames
- function to calculate the luminosity of a hex colour
- function to create a radio station of 25 songs with genre equal to the genre of the selected track
- function to quickly rate songs with a rating provided by the skin - this is used to overlay a binary rating system onto the current 1-10 scale user ratings for songs. So you set a threshold in skin settings e.g. 7, then you can choose to 'like' the song, i.e. give it the rating of the threshold, via this script, which is executed a) from the contextmenu and b) from the music player. 
- function to shuffle all songs by a given artist

For plugins:
- next episode and in progress movies/tv are both provided by script.embuary.helper. But I updated to remove a lot of the warnings/errors in Nexus and switch to InfoTagVideo instead of ListItem.SetInfo. It also supports some infotags introduced in Nexus that I think haven't been added to script.embuary.helper e.g. ListItem.Property(WatchedEpisodePercent). I also stripped out a lot that wasn't required for my use case (e.g. tags, that the documentation said can slow down the json calls significantly) so it should run quicker as well.
- get director - I expanded this widget to incorporate musicvideos as well as movies

For background service:
- a clearlogo cropper service, similar to plugin.video.themovieddb.helper. However it is different in a few fundamental ways, namely that it runs for several custom art types I support in the skin (clearlogo, clearlogo-alt, clearlogo-billboard to have the flexibility of light clearlogo on dark backgrounds, dark on light and a white clearlogo ontop of other artwork). It also outputs luminosity information that I use to determine whether there is enough contrast with the given background for my flags to be coloured with the dominant colour of the cropped image.
- A background fanart service. Unlike what is found in script.embuary.helper, mine only runs when the artwork will be displayed and should also work even if textures haven't been cached.

In addition to the extra functionality I required, the aim with this particular addon is to be as lightweight as possible so it is also much smaller than the two you mentioned and, as a result, I believe it removes the errors/warnings seen in certain scenarios with those other addons.

**Comment by realcopacetic** [comment]
Sorry I just identified an error situation, that's now been resolved, but I submitted a second commit to this pull request. Hope that's ok

**Review by romanvm — CHANGES_REQUESTED** [review]
Please fix errors found by the addon checker

**Comment by realcopacetic** [comment]
Apologies @romanvm - I believe I've made all the changes that were flagged. Thanks.

**Comment by realcopacetic** [comment]
Removed __pychache files and squashed commits following Kodi Addon-Checker second run @romanvm

File: actions.py
Comment (romanvm) [inline actions.py:1]: please remove this file, it is submitted to the root of the repository.

File: addon.xml
Comment (romanvm) [inline addon.xml:1]: please remove this file, it is submitted to the root of the repository.

**Review by romanvm — CHANGES_REQUESTED** [review]

**Comment by romanvm** [comment]
I haven't noticed this at first but please fix the structure of the pull request. Some files are submitted to the root of the target repository.

**Comment by realcopacetic** [comment]
> I haven't noticed this at first but please fix the structure of the pull request. Some files are submitted to the root of the target repository.

Ahh sorry about that, I'm not sure how I did that. I've removed those extraneous files.

**Review by romanvm — CHANGES_REQUESTED** [review]
Please fix addon.xml

**Comment by realcopacetic** [comment]
@romanvm Hi Roman, I've made the latest changes you required. Thanks

**Comment by realcopacetic** [comment]
hi @romanvm - do you think there is anything else holding this one back? What's the approval process like from here? Not trying to pressure you, the only reason I'm asking is because the skin this is for was set live in the Nexus repo a couple of weeks ago, but right now it won't install because of this dependency and the other one (resource.fonts.copacetic). So it would be good to get these live as soon as possible and I have a bit of time over the next few days if further changes are required, but then I'll be starting a new job so may not have much time.
Thanks

File: script.copacetic.helper/plugin.py
Comment (enen92) [inline script.copacetic.helper/plugin.py:6]: dummy init python files are no longer needed in python3

**Review by enen92 — COMMENTED** [review]
LGTM codewise.
Apart from the license nitpick I have a few concerns regarding the fanart the addon is bundled. It's highly recommended to avoid copyrighted imagery. Probably since this is a simple helper script, you can come up with some other alternative that uses the addon icon or similar?

File: script.copacetic.helper/plugin.py
Comment (realcopacetic) [reply script.copacetic.helper/plugin.py:6]: Thanks @enen92my coding knowledge is fairly limited. Can you point me towards an example of how this should be done in python 3?

Comment (realcopacetic) [reply script.copacetic.helper/plugin.py:6]: Thanks @enen92 my coding knowledge is fairly limited. Can you point me towards an example of how this should be done in python 3?

Comment (enen92) [reply script.copacetic.helper/plugin.py:6]: Just remove the empty `__init__.py` that is below in the GitHub diff

Comment (romanvm) [reply script.copacetic.helper/plugin.py:6]: I'd rather disagree. While Python 3 supports so called implicit namespace packages, for regular cases it's better to use old style packages with `__init__.py`.

**Comment by realcopacetic** [comment]
> LGTM codewise. Apart from the license nitpick I have a few concerns regarding the fanart the addon is bundling. It's highly recommended to avoid copyrighted imagery. Probably since this is a simple helper script, you can come up with some other alternative that uses the addon icon or similar?

Hi @enen92 I've replaced fanart and updated licence. Thanks

**Comment by realcopacetic** [comment]
Hi @romanvm @enen92 - what's left for this one before it can be merged? Sorry to keep chasing, I'm just aware of the fact that until this one is in the repo, the skin that depends on it is also DOA in the official skins repo.

**Review by enen92 — APPROVED** [review]

**Comment by enen92** [comment]
Nothing, we just don't have many people around to take care of the plethora of tasks Kodi requires :)

**Comment by realcopacetic** [comment]
> Nothing, we just don't have many people around to take care of the plethora of tasks Kodi requires :)

Thanks, really appreciate both of your help in getting this through the process

**Review by romanvm — COMMENTED** [review]
Ups

**Pull request merged**
